doc/ACL_pytest.md

```diff
@@ -0,0 +1,326 @@
+- *acstests* simply a symbolic link to acs PTF test scripts
+
+
+#### test_acl.py
+
+The current test is divided into 4 cases:
+- Basic test - setup ACL rules, run traffic, teardown ACL rules
+- Incremental test - setup ACL rules incrementaly, run traffic, teardown ACL rules
+- Reboot test - setup ACL rules, reboot, run traffic, teardown ACL rules
+- Port toggle test - setup ACL rules, toggle all ports, run traffic, teardown ACL rules
+
+The test logic will be seperated into generic traffic test in *BaseAclTest* and derivatives are meant to provide fixture to setup/teardown according to their requirements.
+
+According to above cases 4 test classes has to be coded:
+
+- *TestAcl*
+- *TestIncrementalAcl*
+- *TestAclWithReboot*
+- *TestAclWithPortToggle*
+
+
+In common for all above cases we need to do the following setup/teardown:
+- setup test variables according to topology and minigraph facts
+e.g. get the list of ToR/Spine ports, setup names and configuration for ACL tables and so on
+- generate and deploy configuration files to the DUT, after the test remove configuration and cleanup
+
+Also configuration generation and dependent traffic run tests will be parametrized for 'ingress', 'egress' ACL tests
+
+##### Test fixtures
+
+As above described setup and configuration generation can be done in scope of *test_acl.py* module only once.
+
+These fixtures, by the way,can live in *acl/conftest.py*, so other ACL test modules can reuse them.
+
+###### Module scope
+
+```python
+@pytest.fixture(scope='module')
+def setup(dut, testbed):
+    '''
+    Setup fixture gathers all test required information from DUT facts and testbed
+    '''
+
+@pytest.fixture(scope='module', params=['ingress', 'egress'])
+def stage(request):
+    '''
+    Small fixture to return the ACL table stage is beeing tested
+    '''
+
+@pytest.yield_fixture(scope='module')
+def config(request, dut, setup):
+    '''
+    Fixture generates ACL tables, ACL rules configuration and deploys on the DUT
+    The return value is config files generated
+    '''
+
+@pytest.yield_fixture(scope='module')
+def table(dut, config, stage):
+    '''
+    Create an ACL table on DUT
+    '''
+
+```
+
+Based on the above it is assumed that tests in one module will reuse the ACL table created by 'table' fixture.
+
+If, for example, some one is developing ACL table scale test he will need a seperate test module ('test_acl_scale.py') and override 'config' fixutre to generate many tables configuration and reuse 'table' fixture.
+
+###### Class scope
+
+
+```python
+
+class BaseAclTest(object):
+    __meta__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def setup_rules(self):
+        pass
+
+    @abc.abstractmethod
+    def teardown_rules(self):
+        pass
+
+    @pytest.yield_fixture(scope='class')
+    def rules(self, dut, table):
+        '''
+        Setup/Teardown ACL rules on DUT
+        '''
+
+    @pytest.yield_fixture(scope='class', autouse=True)
```

Comment:
IMO, couters checking is in the middle of a testing. Why not just make it an ordinary method/function so that test cases can call it when needed?

Reply:
You are right, the design badly explains how the counters check work.
I would prefer having it after traffic test cases finished.
If I will call counter check after each traffic test case I will slow down the test a lot. Each counter check has to wait at least 10 sec for ACL counters update.

There is a problem that if I ran single traffic test case how do I pass which rule to check for counters
Still thinking about counters check...